### PR TITLE
OCLOMRS-967:Pick concepts from a dictionary (in addition to sources)

### PR DESCRIPTION
**Pull request by jwnasambu**
# JIRA TICKET NAME:
[Pick concepts from a dictionary (in addition to sources)](https://issues.openmrs.org/browse/OCLOMRS-967)

# Summary:

I worked on a feature that allows a user to create a dictionary that contains concepts from another dictionary.

### CODE REVIEW

File: src/apps/concepts/components/ViewConceptsHeader.tsx [hunk 1]
```diff
@@ -152,6 +159,7 @@ const ViewConceptsHeader: React.FC<Props> = ({
               >
                 <AccountTreeOutlined className={classes.sourceIcon}/> 
                 {preferredSourceName}
+                <FolderOpen  className={classes.dictionaryIcon}/>
```
Comment (ibacher) [inline src/apps/concepts/components/ViewConceptsHeader.tsx:162]:
```suggestion
```

File: src/apps/concepts/components/ViewConceptsHeader.tsx [hunk 2]
```diff
@@ -136,6 +142,7 @@ const ViewConceptsHeader: React.FC<Props> = ({
                 >
                   <AccountTreeOutlined className={classes.sourceIcon}/> 
                   {name}
+                  <FolderOpen  className={classes.dictionaryIcon}/>
```
Comment (ibacher) [inline src/apps/concepts/components/ViewConceptsHeader.tsx:145]:
```suggestion
```

**Review by ibacher — COMMENTED** [review]
So, some quick thoughts on this:

* at line 132 of ViewConceptsHeader remove `.concat(sources)`
* `formatPreferredSources` should be defined like this:

```ts
const [preferredSources, setPreferredSources] = useState();
useEffect(() => {
  sources = Object.entries(PREFERRED_SOURCES_VIEW_ONLY);

  if (showSources) {
    sources.concat(sources).concat(dictionaries);
  }

  setPreferredSources(
    sources.map(([key, value]) => ({ name: key, url: value }));
  );
}, [PREFERRED_SOURCES_VIEW_ONLY, showSources, sources, dictionaries])
```

Then line 132 on should look something like:

```ts
{(preferredSources?.map(
  ({ name, url }) => (
    <MenuItem
      // replace because we want to keep the back button useful
      replace
      to={gimmeAUrl({}, `${url}concepts/`)}
      key={name}
      component={Link}
      onClick={handleSwitchSourceClose}
      data-testid={name}
    >
      { url?.contains("/collection") ? (
        <FolderOpen className={classes.sourceIcon} />
      ) : (
        <AccountTreeOutlined className={classes.sourceIcon} />
      ) }
      { name }
    </MenuItem>
  )
)}
```

That should also eliminate the need for lines 149-164 or so.

File: src/apps/concepts/components/ViewConceptsHeader.tsx [hunk 3]
```diff
@@ -70,13 +70,25 @@ const ViewConceptsHeader: React.FC<Props> = ({
   gimmeAUrl,
   addConceptToDictionary,
   children,
-  sources,
+
 }) => {
   const [showSources, setShowSources] = useState(false);
-  const formatPrefferedSources = Object.entries(PREFERRED_SOURCES_VIEW_ONLY).map(([key, value]) => ({ name: key , url: value }));
+  const [preferredSources, setPreferredSources] = useState();
+useEffect(() => {
```
Comment (ibacher) [inline src/apps/concepts/components/ViewConceptsHeader.tsx:77]: You need to import `useEffect()` if your editor doesn't do this automatically, it goes in line 1 of this file, i.e.,

```ts
import React, { useEffect, useState } from "react";
```

File: src/apps/concepts/components/ViewConceptsHeader.tsx [hunk 4]
```diff
@@ -70,13 +70,25 @@ const ViewConceptsHeader: React.FC<Props> = ({
...
+  sources = Object.entries(PREFERRED_SOURCES_VIEW_ONLY);
+
+  if (showSources) {
+    sources.concat(sources).concat(dictionaries);
+  }
+
+  setPreferredSources(
+    sources.map(([key, value]) => ({ name: key, url: value }));
+  );
+}, [PREFERRED_SOURCES_VIEW_ONLY, showSources, sources, dictionaries])
```
Comment (ibacher) [inline src/apps/concepts/components/ViewConceptsHeader.tsx:87]:
```suggestion
  useEffect(() => {
    sources = Object.entries(PREFERRED_SOURCES_VIEW_ONLY);

    if (showSources) {
      sources.concat(sources).concat(dictionaries);
    }

    setPreferredSources(
      sources.map(([key, value]) => ({ name: key, url: value }));
    );
  }, [PREFERRED_SOURCES_VIEW_ONLY, showSources, sources, dictionaries])
```

Comment (ibacher) [reply src/apps/concepts/components/ViewConceptsHeader.tsx:87]: You'll notice that in this line, we added `dictionaries` as a requirement, but there's no local value named "dictionaries". We need to add that (see next comment).

File: src/apps/concepts/components/ViewConceptsHeader.tsx [hunk 5]
```diff
@@ -70,13 +70,25 @@ const ViewConceptsHeader: React.FC<Props> = ({
   gimmeAUrl,
   addConceptToDictionary,
   children,
-  sources,
+
```
Comment (ibacher) [inline src/apps/concepts/components/ViewConceptsHeader.tsx:73]:
```suggestion
  sources,
  dictionaries
```

File: src/apps/concepts/components/ViewConceptsHeader.tsx [hunk 6]
```diff
@@ -70,13 +70,25 @@ const ViewConceptsHeader: React.FC<Props> = ({
...
+  const [preferredSources, setPreferredSources] = useState();
+useEffect(() => {
+  sources = Object.entries(PREFERRED_SOURCES_VIEW_ONLY);
+
+  if (showSources) {
+    sources.concat(sources).concat(dictionaries);
+  }
+
+  setPreferredSources(
+    sources.map(([key, value]) => ({ name: key, url: value }));
```
Comment (ibacher) [inline src/apps/concepts/components/ViewConceptsHeader.tsx:85]:
```suggestion
    sources.map(([key, value]) => ({ name: key, url: value }))
```

**Review by ibacher — COMMENTED** [review]
Let's see where these changes get us.

**Comment by jwnasambu** [comment]
@ibacher Thanks for the review.  Kindly let me make the proposed changed.

File: src/apps/concepts/pages/ViewConceptsPage.tsx [hunk 7]
```diff
@@ -111,13 +117,15 @@ const ViewConceptsPage: React.FC<Props> = ({
...
   dictionary,
+  dictionaries = [],
   source,
   retrievePublicSources,
   sources = [],
   loading,
   errors,
   retrieveConcepts,
   retrieveDictionary,
+  retrievePublicDictionaries,
```
Comment (hadijahkyampeire) [inline src/apps/concepts/pages/ViewConceptsPage.tsx:128]: We need to fetch dictionaries on this page because I am not seeing them.

**Review by hadijahkyampeire — COMMENTED** [review]
@jwnasambu @ibacher I am going to push some more changes about what I have commented on.

File: src/apps/concepts/components/ViewConceptsHeader.tsx [hunk 8]
```diff
@@ -64,10 +66,26 @@ const ViewConceptsHeader: React.FC<Props> = ({
...
+  >();
+  useEffect(() => {
+    const defaultSources = Object.entries(
+      PREFERRED_SOURCES_VIEW_ONLY
+    ).map(([key, value]) => ({ name: key, url: value }));
+
+    if (showSources) {
+      defaultSources
+        .concat(sources.map(s => ({ name: s.name, url: s.url })))
+        .concat(dictionaries.map(d => ({ name: d.name, url: d.url })));
```
[the remark this answers is not in the document]
Comment (ibacher) [reply src/apps/concepts/components/ViewConceptsHeader.tsx:84]: Ok... I was under the impression that we want to keep the default sources list as is and only change things when the user is clicking the "Select other source", but, sure

**Comment by hadijahkyampeire** [comment]
@jwnasambu I have pushed some small changes to retrieve the dictionaries when you fetch the latest changes you should be able to see dictionaries, Also the list was not showing the public concepts so I fixed that in the useEffect. cc @ibacher .
Screenshot 2021-06-16 at 22 34 04


There are some visual changes I remember people suggesting on one of the OCL call, but I don't remember them well, @jwnasambu  you might be remembering them. One thing that I remember was they talked something about changing the `Public Sources` word to some other word but I don't remember.

**Comment by ibacher** [comment]
@hadijahkyampeire Cool! "Public Sources"... maybe "All Public Concepts"?

Also, I know this isn't *this* ticket, but we really need to filter the list of sources there so that they are only sources that can be used in an OpenMRS Dictionary.

**Comment by hadijahkyampeire** [comment]
> @hadijahkyampeire Cool! "Public Sources"... maybe "All Public Concepts"?
> 
> Also, I know this isn't _this_ ticket, but we really need to filter the list of sources there so that they are only sources that can be used in an OpenMRS Dictionary.

@ibacher, Maybe we can make a follow-up ticket and @jwnasambu can work on it after we merge this, I agree the list is too long.

**Comment by hadijahkyampeire** [comment]
@jwnasambu why do we have two PRS for this work, can you please close one and maintain one?

**Pull request closed without merging**